### DISCUSSION TITLE
maximum tone() frequency?

### DISCUSSION THREAD

**Discussion by stingpie** [earlephilhower/arduino-pico#301] [Q&A] [answered]
I want to see whether I can use a raspberry pi pico as a frequency generator for some ultrasonic experiments I have planned. However, I need to get around the 200 mhz range, so I'm wondering just how high a frequency the tone() function can output.

**Comment by earlephilhower** [comment, marked as answer]
> 200 MHz

I assume you mean kHZ, not MHz :). 

The PIO hardware should be able to do up to 125MHZ / 8 = 20MHz (because each loop uses 8 PIO insns).

However, the code in https://github.com/earlephilhower/arduino-pico/blob/a8121f31d4b69b920502e1ef7de0d588f5e644b4/cores/rp2040/Tone.cpp#L66-L68 does limit the frequency the API allows to 100kHZ.  If you comment that out you should be able to go to 1/2us = 500kHZ.

**Reply by stingpie** [reply]
Thank you, but I really did mean megahertz. I'm doing an experiment to see the effects of ultrasound on yeast, and 211 megahertz is the exact frequency to give the wavelength of a yeast cell. I'm going to have to figure out how to generate that frequency some other way. Do you think a teensy 4 would be able to? I know it has a 600 mhz clock frequency.

**Reply by earlephilhower** [reply]
I would imagine no microcontroller will generate that kind of frequency on a pin under SW control, sorry.  GPIO rise/fall slew rates plus the fact that many IO rings run at a peripheral clock which is often fixed and some fraction of the CPU's.

I'm also not sure how you would get a physical transducer that would function at that range...even medical ultrasound is <20MHz (and it is generated by special-purpose HW and the transducer is custom-built).

You'd also end up radiating radio waves in the 211MHz FM/maritime bands and all the associated harmonics (since you're driving a square wave you get all those for free).

Making a 211MHz wave is simple enough w/proper transistor design, just look at schematics for a tunable radio transmitter.  You could then gate it on/off w/a GPIO pin at low rates (i.e. for duty cycle) w/o worrying about MHz-range signals.
